core/include/traccc/fitting/kalman_filter/kalman_fitter.hpp

```diff
@@ -0,0 +1,202 @@
+
+                filter(new_seed_params, fitter_state);
+            }
+        }
+    }
+
+    /// Run the kalman fitter for an iteration
+    ///
+    /// @tparam seed_parameters_t the type of seed track parameter
+    ///
+    /// @param seed_params seed track parameter
+    /// @param fitter_state the state of kalman fitter
+    template <typename seed_parameters_t>
+    void filter(const seed_parameters_t& seed_params, state& fitter_state) {
+
+        // Create propagator
+        propagator_type propagator({}, {});
+
+        // Set path limit
+        fitter_state.m_aborter_state.set_path_limit(m_cfg.pathlimit);
+
+        // Create propagator state
+        typename propagator_type::state propagation(
+            seed_params, m_detector->get_bfield(), *m_detector);
+
+        // Set overstep tolerance and stepper constraint
+        propagation._stepping().set_overstep_tolerance(
+            m_cfg.overstep_tolerance);
+        propagation._stepping
+            .template set_constraint<detray::step::constraint::e_accuracy>(
+                m_cfg.step_constraint);
+
+        // Run forward filtering
+        propagator.propagate(propagation, fitter_state());
+
+        // Run smoothing
+        smooth(fitter_state);
+
+        //@todo: Write track info
+    }
+
+    /// Run smoothing after kalman filtering
+    ///
+    /// @brief The smoother is based on "Application of Kalman filtering to
+    /// track and vertex fitting", R.Frühwirth, NIM A.
+    ///
+    /// @param fitter_state the state of kalman fitter
+    void smooth(state& fitter_state) {
+        auto& track_states = fitter_state.m_fit_actor_state.m_track_states;
+
+        // The smoothing algorithm requires the following:
+        // (1) the filtered track parameter of the current surface
+        // (2) the smoothed track parameter of the next surface
+        //
+        // Since the smoothed track parameter of the last surface can be
+        // considered to be the filtered one, we can reversly iterate the
+        // algorithm to obtain the smoothed parameter of other surfaces
+        auto& last = track_states.back();
+        last.smoothed().set_vector(last.filtered().vector());
+        last.smoothed().set_covariance(last.filtered().covariance());
+
+        const auto& mask_store = m_detector->mask_store();
+
+        for (typename vecmem::vector<
+                 track_state<transform3_type>>::reverse_iterator it =
+                 track_states.rbegin() + 1;
+             it != track_states.rend(); ++it) {
+
+            // Surface
+            const auto& surface =
+                m_detector->surface_by_index(it->surface_link());
+
+            // Run kalman smoother
+            mask_store.template call<gain_matrix_smoother<transform3_type>>(
+                surface.mask(), *it, *(it - 1));
+        }
+    }
+
+    private:
+    // Detector object
+    const detector_type* m_detector;
```

Comment:
Let's make things fairly technical here. :stuck_out_tongue:
  - This way of holding onto a pointer to the detector is not a bad one. I myself used to prefer this way of coding as well.
  - But what I learned since then is that in modern C\+\+ one should instead choose between the following 2 options instead:
    * If the type (in this case `traccc::kalman_fitter`) is meant to be copy-able, it's slightly more elegant to use the [std::reference_wrapper](https://en.cppreference.com/w/cpp/utility/functional/reference_wrapper) type to hold on to the reference/pointer. That way it is very clear in the code that this is **always** meant to point at something valid. Whereas with a pointer you may need to add some lines of documentation to clarify this.
    * If the type is not meant to be copy-able, then one should just use a reference member variable. I.e.

```c++
const detector_type& m_detector;
```

Since I think for this particular class (and the others that use `detector` as well) copying is a valid operation, [std::reference_wrapper](https://en.cppreference.com/w/cpp/utility/functional/reference_wrapper) could be an elegant thing to make use of.

But as I started, this is definitely not something mandatory. The pointer works functionally exactly the same. :wink:

Reply:
> That way it is very clear in the code that this is always meant to point at something valid.

Is it like `std::reference_wrapper` won't be compiled if the wrapped type is not copy-able?
Anyway I changed to `std::reference_wrapper` as I don't have any objection on this.

Reply:
Ah OK, it doesn't allow the null value